src/MongoDB.Driver.Core/Core/Authentication/ScramShaAuthenticator.cs

```diff
@@ -189,22 +194,77 @@ private class ClientFirst : ISaslStep
                 _hi = hi;
                 _hmac = hmac;
                 _rPrefix = rPrefix;
+                _originalThreadId = Thread.CurrentThread.ManagedThreadId;
+                _originalCulture = CultureInfo.CurrentCulture.ToString();
             }
 
             public byte[] BytesToSendToServer => _bytesToSendToServer;
 
             public bool IsComplete => false;
 
+            private string BuildErrorMessage(
+                string baseMessage,
+                string connectionId,
+                byte[] bytesToSendToServer,
+                byte[] bytesReceivedFromServer,
+                string serverFirstMessage,
+                string rPrefix,
+                string r,
+                string originalThreadId,
+                string currentThreadId,
+                string originalCulture,
+                string currentCulture,
+                string clientFirstMessageBare)
+            {
+                var builder = new StringBuilder();
+
+                builder.AppendLine("Base message:" + baseMessage);
+                builder.AppendLine($"{nameof(connectionId)}: {connectionId}");
+
+                builder.AppendLine($"Hex_{nameof(bytesToSendToServer)}: {BsonUtils.ToHexString(bytesToSendToServer)}");
+                builder.AppendLine($"Hex_{nameof(bytesReceivedFromServer)}: {BsonUtils.ToHexString(bytesReceivedFromServer)}");
+
+                builder.AppendLine($"{nameof(rPrefix)}: {rPrefix}");
+                builder.AppendLine($"{nameof(r)}: {r}");
+                builder.AppendLine($"{nameof(serverFirstMessage)}: {serverFirstMessage}");
+                builder.AppendLine($"{nameof(originalThreadId)}: {originalThreadId}");
+                builder.AppendLine($"{nameof(currentThreadId)}: {currentThreadId}");
+                builder.AppendLine($"{nameof(originalCulture)}: {originalCulture}");
+                builder.AppendLine($"{nameof(currentCulture)}: {currentCulture}");
+
+                builder.AppendLine($"{nameof(clientFirstMessageBare)}: {clientFirstMessageBare}");
+                builder.AppendLine($"{nameof(bytesToSendToServer)}: {string.Join(",", bytesToSendToServer)}");
+                builder.AppendLine($"{nameof(bytesReceivedFromServer)}: {string.Join(",", bytesReceivedFromServer)}");
```

Comment:
No need to include the same data twice.

The hex encoding is sufficient.

Reply:
done

Reply:
done